stock_cancel/models/stock_picking.py

```diff
@@ -0,0 +1,59 @@
+# -*- coding: utf-8 -*-
+# Copyright 2012 Andrea Cometa
+# Copyright 2013 Agile Business Group sagl
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl.html).
+
+from odoo import api, models, exceptions, _
+
+
+class StockPicking(models.Model):
+    _inherit = 'stock.picking'
+
+    @api.multi
+    def cancel_valuation_moves(self):
+        self.ensure_one()
+        account_moves = self.env['account.move'].search(
+            [('ref', '=', self.name)])
+        account_moves.button_cancel()
+        return account_moves
+
+    def _check_restrictions(self):
+        # returned_move_ids in stock.move
+        # split_from in stock.move
+        if self.search([('backorder_id', '=', self.id)], limit=1):
+            raise exceptions.UserError(
+                _('Not Allowed, picking has backorder.'))  # noqa
+
+    @api.multi
+    def action_revert_done(self):
+        for picking in self:
+            picking._check_restrictions()
+            acc_moves = picking.cancel_valuation_moves()
+            acc_moves.unlink()
+            if picking.invoice_id.filtered(
+                    lambda order: order.state != 'cancel'):
+                raise exceptions.UserError(
+                    _('Picking %s has invoices') % (picking.name))
+            picking.move_lines.write({'state': 'draft'})
+            # reassign quants done
+            for move in picking.move_lines:
+                move._check_restrictions()
+                move.quant_ids._revert()
+
+            picking.group_id.procurement_ids.reset_to_confirmed()
+            sale_lines = picking.group_id.procurement_ids.mapped(
+                'sale_line_id')
+            for line in sale_lines:
+                line.qty_delivered = line._get_delivered_qty()
+            if sale_lines:
+                order = sale_lines[0].order_id
+                delivery_line = order.with_context(
```

Comment:
Hi,

In a installation without delivery module installed. If your try to re-open a picking that comes from a sale, you'll get this error:

`AttributeError: 'sale.order.line' object has no attribute 'is_delivery'`

You should move this piece of code to stock_cancel_delivery module.

Reply:
I see. I have to move that to the other module proposed, stock_cancel_delivery. Thanks